### PR TITLE
gh-125451: Fix deadlock in ProcessPoolExecutor shutdown

### PR DESCRIPTION
**Pull request by colesbury**
There was a deadlock when `ProcessPoolExecutor` shuts down at the same time that a queueing thread handles an error when processing a task. Don't use `_shutdown_lock` to protect the `_ThreadWakeup` pipes -- use an internal lock instead. This fixes the ordering deadlock where the `ExecutorManagerThread` holds the `_shutdown_lock` and joins the queueing thread, while the queueing thread is attempting to acquire the `_shutdown_lock` while closing the `_ThreadWakeup`.

* Issue: gh-125451

### CODE REVIEW

File: Lib/concurrent/futures/process.py [hunk 1]
```diff
@@ -68,27 +68,30 @@
...
-            self._writer.send_bytes(b"")
+        with self._lock:
+            if not self._closed:
+                self._writer.send_bytes(b"")
 
     def clear(self):
-        if not self._closed:
-            while self._reader.poll():
-                self._reader.recv_bytes()
+        assert not self._closed
```
Comment (vstinner) [inline Lib/concurrent/futures/process.py]: Maybe replace the assertion with an `if`? You can raise `RuntimeError("operation on closed _ThreadWakeup")` or something like that.

**Review by vstinner — COMMENTED** [review]
Near line 724, please remove the comment:

```
        # _shutdown_lock must be locked to access _ThreadWakeup (...)
```

---

Oh, I made commit a4dfe8ede5a37576e17035dccfe109ba7752237e in 2020 to fix another race condition:

```
commit a4dfe8ede5a37576e17035dccfe109ba7752237e
Author: Victor Stinner <vstinner@python.org>
Date:   Wed Apr 29 03:32:06 2020 +0200

    bpo-39995: Fix concurrent.futures _ThreadWakeup (GH-19760)
    
    Fix a race condition in concurrent.futures._ThreadWakeup: access to
    _ThreadWakeup is now protected with the shutdown lock.
```

**Comment by vstinner** [comment]
> buildbot/x86 Debian Installed with X PR — Build done.
> buildbot/x86 Debian Non-Debug with X PR — Build done. 

The failure of these 2 workers is known: https://github.com/python/cpython/issues/125535

File: Lib/concurrent/futures/process.py (hunk 1 above)
Comment (kumaraditya303) [inline Lib/concurrent/futures/process.py]:
```suggestion
        if not self._closed:
            raise RuntimeError('operation on closed _ThreadWakeup')
```

**Review by vstinner — APPROVED** [review]
LGTM.

I confirm that this change fix https://github.com/python/cpython/issues/125451 issue.

Using a lock per _ThreadWakeup, instead of reusing the shutdown lock, sounds like a good idea.

**Review by kumaraditya303 — APPROVED** [review]

**Pull request merged**